python/ray/dag/tests/experimental/test_execution_schedule.py

```diff
@@ -226,6 +231,174 @@ def test_two_nccl_writes(self, monkeypatch):
+        The following test case simulates a scenario where the READ
+        operation for every all-reduce input has already been added to the
+        execution schedule.
+        """
+        monkeypatch.setattr(ActorHandle, "__init__", mock_actor_handle_init)
+        fake_actor_1, dag_idx_1, local_idx_1 = ActorHandle("fake_actor_1"), 1, 0
+        fake_actor_2, dag_idx_2, local_idx_2 = ActorHandle("fake_actor_2"), 2, 0
+        dag_idx_3, local_idx_3 = 3, 1
+        mock_graph = {
+            dag_idx_1: generate_dag_graph_nodes(
+                local_idx_1, dag_idx_1, fake_actor_1, False, is_collective=True
+            ),
+            dag_idx_2: generate_dag_graph_nodes(
+                local_idx_2, dag_idx_2, fake_actor_2, False, is_collective=True
+            ),
+            dag_idx_3: generate_dag_graph_nodes(
+                local_idx_3, dag_idx_3, fake_actor_2, False, is_collective=True
+            ),
+        }
+        del mock_graph[dag_idx_1][_DAGNodeOperationType.READ]
+        del mock_graph[dag_idx_2][_DAGNodeOperationType.READ]
+        del mock_graph[dag_idx_3][_DAGNodeOperationType.READ]
+
+        _add_edge(
+            mock_graph[dag_idx_1][_DAGNodeOperationType.COMPUTE],
+            mock_graph[dag_idx_1][_DAGNodeOperationType.WRITE],
+        )
+        _add_edge(
+            mock_graph[dag_idx_2][_DAGNodeOperationType.COMPUTE],
+            mock_graph[dag_idx_2][_DAGNodeOperationType.WRITE],
+        )
+        _add_edge(
+            mock_graph[dag_idx_3][_DAGNodeOperationType.COMPUTE],
+            mock_graph[dag_idx_3][_DAGNodeOperationType.WRITE],
+        )
+        mock_actor_to_candidates = {
+            fake_actor_1: [mock_graph[dag_idx_1][_DAGNodeOperationType.COMPUTE]],
+            fake_actor_2: [
+                mock_graph[dag_idx_2][_DAGNodeOperationType.COMPUTE],
+                mock_graph[dag_idx_3][_DAGNodeOperationType.COMPUTE],
+            ],
+        }
+        # Embed collective group information in _DAGNodeOperation.
+        collective_group_1 = {
+            (dag_idx_1, _DAGNodeOperationType.COMPUTE),
+            (dag_idx_2, _DAGNodeOperationType.COMPUTE),
+        }
+        mock_graph[dag_idx_1][
+            _DAGNodeOperationType.COMPUTE
+        ].collective_group = collective_group_1
+        mock_graph[dag_idx_2][
+            _DAGNodeOperationType.COMPUTE
+        ].collective_group = collective_group_1
+        collective_group_2 = {
+            (dag_idx_3, _DAGNodeOperationType.COMPUTE),
+        }
+        mock_graph[dag_idx_3][
+            _DAGNodeOperationType.COMPUTE
+        ].collective_group = collective_group_2
+        next_nodes = _select_next_nodes(mock_actor_to_candidates, mock_graph)
+        assert len(next_nodes) == 2
+        assert set(next_nodes) == {
+            mock_graph[dag_idx_1][_DAGNodeOperationType.COMPUTE],
+            mock_graph[dag_idx_2][_DAGNodeOperationType.COMPUTE],
+        }
```

Comment:
Check that the second call to `_select_next_nodes` returns the other collective?

Reply:
Done.